File: src/main/scala/org/xbery/artbeams/users/service/UserSubscriptionService.scala
package org.xbery.artbeams.users.service

import java.time.Instant

/**
  * @author Radek Beran
  */
trait UserSubscriptionService {
  /**
    * Subscribe user with given full name and email to newsletters and send him/her a product
    * (as a reward).
    * @param fullName
    * @param email
    * @param productId
    */
  def subscribe(fullName: Option[String], email: String, productId: String): Unit

  /**
    * Confirms user's consent with personal data processing and sending of newsletters.
    * @param fullName full name of user
    * @param email email of user
    * @param productId product ordered with the registration to newsletter
    * @return time of consent confirmation if consent was successfully confirmed and stored
    */
  def confirmConsent(fullName: Option[String], email: String, productId: String): Option[Instant]
}

File: src/main/kotlin/org/xbery/artbeams/users/service/UserSubscriptionService.kt
package org.xbery.artbeams.users.service

import org.xbery.artbeams.orders.domain.OrderState
import org.xbery.artbeams.products.domain.Product
import org.xbery.artbeams.users.domain.User

/**
 * @author Radek Beran
 */
interface UserSubscriptionService {

    /**
     * Confirms user's consent with personal data processing and sending of newsletters.
     * This may lead to new registration of user if the user with given e-mail = login does not exist yet.
     * This leads to creation of new order of given product for the user.
     *
     * @param fullName full name of user
     * @param login login = email of user
     * @param product product ordered with the registration to newsletter
     * @return user created or updated
     */
    fun createOrUpdateUserWithOrderAndConsent(
        fullName: String?,
        login: String,
        product: Product,
        orderNumber: String,
        orderState: OrderState
    ): User
}
